bioio_ome_zarr/writers/ome_zarr_writer.py

```diff
@@ -0,0 +1,448 @@
+from typing import Any, Dict, List, Literal, Optional, Tuple, Union, cast
+
+import dask.array as da
+import numcodecs
+import numpy as np
+import zarr
+from numcodecs import Blosc as BloscV2
+from zarr.codecs import BloscCodec, BloscShuffle
+
+from .metadata import Axes, Channel, MetadataParams, build_ngff_metadata
+from .utils import (
+    chunk_size_from_memory_target,
+    compute_level_chunk_sizes_zslice,
+    resize,
+)
+
+ChunkSingle = Tuple[int, ...]
+ChunkPerLevel = Tuple[ChunkSingle, ...]
+ChunkShape = Union[ChunkSingle, ChunkPerLevel]
+
+
+class OMEZarrWriter:
+    """
+    OMEZarrWriter is a unified OME-Zarr writer that targets either Zarr v2
+    (NGFF 0.4) or Zarr v3 (NGFF 0.5) with the same public API. Supports
+    2 ≤ N ≤ 5 dimensions (e.g., YX, ZYX, TYX, CZYX, or TCZYX) and writes a
+    multiscale pyramid with nearest-neighbor downsampling.
+    """
+
+    def __init__(
+        self,
+        store: Union[str, zarr.storage.StoreLike],
+        shape: Tuple[int, ...],
+        dtype: Union[np.dtype, str],
+        *,
+        scale: Optional[Tuple[Tuple[float, ...], ...]] = None,
+        chunk_shape: Optional[ChunkShape] = None,
+        shard_factor: Optional[Tuple[int, ...]] = None,
+        compressor: Optional[Union[BloscCodec, numcodecs.abc.Codec]] = None,
+        zarr_format: Literal[2, 3] = 3,
+        image_name: Optional[str] = "Image",
+        channels: Optional[List[Channel]] = None,
+        rdefs: Optional[dict] = None,
+        creator_info: Optional[dict] = None,
+        root_transform: Optional[Dict[str, Any]] = None,
+        axes_names: Optional[List[str]] = None,
+        axes_types: Optional[List[str]] = None,
+        axes_units: Optional[List[Optional[str]]] = None,
+        physical_pixel_size: Optional[List[float]] = None,
+    ) -> None:
+        """
+        Initialize the writer and capture core configuration. Arrays and
+        metadata are created lazily on the first write. Does not write to
+        disk until data is written.
+
+        Parameters
+        ----------
+        store : Union[str, zarr.storage.StoreLike]
+            Filesystem path, URL (via fsspec), or Store-like for the root
+            group.
+        shape : Tuple[int, ...]
+            Level-0 image shape (e.g., (T,C,Z,Y,X)).
+        dtype : Union[np.dtype, str]
+            NumPy dtype for the on-disk array.
+        scale : Optional[Tuple[Tuple[float, ...], ...]]
```

Comment:
I still have a weird bad feeling about passing floating-point numbers in here instead of explicit level shapes.  It just seems less precise and harder to debug vs having the caller figure out what the level shapes should be and send them in.

Reply:
we could allow for both, ints for scale factors, floats for percentages